lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,382 @@
+    }
+
+    return null;
+  }
+
+  /// Returns the path to one subpath variant, resolved in the `exports` of a
+  /// package manifest.
+  ///
+  /// Throws an error if multiple `subpathVariants` match, and null if none
+  /// match.
+  ///
+  /// Implementation of `PACKAGE_EXPORTS_RESOLVE` from the [Resolution Algorithm
+  /// Specification](https://nodejs.org/api/esm.html#resolution-algorithm-specification).
+  String? _nodePackageExportsResolve(
+      String packageRoot,
+      List<String?> subpathVariants,
+      Object exports,
+      String? subpath,
+      String packageName) {
+    if (exports is Map<String, dynamic> &&
+        exports.keys.any((key) => key.startsWith('.')) &&
+        exports.keys.any((key) => !key.startsWith('.'))) {
+      throw '`exports` in $packageName can not have both conditions and paths '
+          'at the same level.\n'
+          'Found ${exports.keys.map((key) => '"$key"').join(',')} in '
+          '${p.join(packageRoot, 'package.json')}.';
+    }
+    String? processVariant(String? variant) {
+      if (variant == null) {
+        return _getMainExport(exports).andThen((mainExport) =>
+            _packageTargetResolve(variant, mainExport, packageRoot));
+      }
+      if (exports is! Map<String, dynamic> ||
+          exports.keys.every((key) => !key.startsWith('.'))) {
+        return null;
+      }
+      var matchKey = "./${p.toUri(variant)}";
+      if (exports.containsKey(matchKey) &&
+          exports[matchKey] != null &&
+          !matchKey.contains('*')) {
+        return _packageTargetResolve(
+            matchKey, exports[matchKey] as Object, packageRoot);
+      }
+
+      var expansionKeys = [
+        for (var key in exports.keys)
+          if ('*'.allMatches(key).length == 1) key
+      ]..sort(_compareExpansionKeys);
+
+      for (var expansionKey in expansionKeys) {
+        var [patternBase, patternTrailer] = expansionKey.split('*');
+        if (!matchKey.startsWith(patternBase)) continue;
+        if (matchKey == patternBase) continue;
+        if (patternTrailer.isEmpty ||
+            (matchKey.endsWith(patternTrailer) &&
+                matchKey.length >= expansionKey.length)) {
+          var target = exports[expansionKey] as Object?;
+          if (target == null) continue;
+          var patternMatch = matchKey.substring(
+              patternBase.length, matchKey.length - patternTrailer.length);
+          return _packageTargetResolve(
+              variant, target, packageRoot, patternMatch);
+        }
+      }
+
+      return null;
+    }
+
+    var matches = subpathVariants.map(processVariant).whereNotNull().toList();
+
+    return switch (matches) {
+      [var path] => path,
+      [_, _, ...] && var paths =>
+        throw "Unable to determine which of multiple potential resolutions "
+            "found for ${subpath ?? 'root'} in $packageName should be used. "
+            "\n\nFound:\n"
+            "${paths.join('\n')}",
+      _ => null
```

Comment:
It would be simpler to reorder these:

```suggestion
      [] => null,
      var paths =>
        throw "Unable to determine which of multiple potential resolutions "
            "found for ${subpath ?? 'root'} in $packageName should be used. "
            "\n\nFound:\n"
            "${paths.join('\n')}"
```

...since this guarantees that by the time you hit the last case there's more than one path.